repl/src/main/scala/ammonite/repl/interp/Interpreter.scala

```diff
@@ -454,16 +493,65 @@ class Interpreter(prompt0: Ref[String],
 
       def exec(file: Path): Unit = apply(read(file))
 
-      def module(file: Path): Unit = {
+      def loadModule(file: Path, storage: Storage, cacheTag: String): Unit ={
+        val code = read(file)
         val (pkg, wrapper) = Util.pathToPackageWrapper(file, wd)
-        processModule(read(file), wrapper, pkg) match{
+        processModule(code, wrapper, pkg) match {
           case Res.Failure(ex, s) => throw new CompilationError(s)
           case Res.Exception(t, s) => throw t
-          case x => //println(x)
+          case Res.Success(data) =>
+            if(scriptCaching) {
+              val files = for (d <- data._2) yield (d._1, d._2)
+              storage.asInstanceOf[Storage.Folder]classFilesListSave(pkg, wrapper, files, cacheTag)
+            }
         }
         init()
       }
 
+      // Try loading script from cache
+      def cachedModule(path: Path) = {
+        Timer.startTime = System.nanoTime()
+
+        val code: String = read(path)
+        val (pkg, wrapper) = Util.pathToPackageWrapper(path, wd)
+        val cacheTag = "cache" + Util.md5Hash(Iterator(code.getBytes))
+          .map("%02x".format(_)).mkString
+        if(!code.contains("load(")) {
+          storage.asInstanceOf[Storage.Folder].classFilesListLoad(pkg, wrapper, cacheTag) match {
+            case Seq() =>
+              loadModule(path, storage.asInstanceOf[Storage.Folder], cacheTag)
+            case cachedData =>
+              def evalMain(cls: Class[_]) =
+                cls.getDeclaredMethod("$main").invoke(null)
+
+              var blockNumber = 1
+              def getBlockNumber = blockNumber match {
+                case 1 => ""
+                case _ => "_" + blockNumber.toString
+              }
+              cachedData.foreach { d => {
+                for {
+                  cls <- interp.eval.loadClass(pkg + "." + wrapper + getBlockNumber, d._1)
+                } yield {
+                  evalMain(cls)
+                }
+                blockNumber += 1
+              }
+              }
+          }
+        } else loadModule(path, storage.asInstanceOf[Storage.Folder], cacheTag)
+      }
+
+      def module(file: Path): Unit = {
+        if (!withCompiler)
```

Comment:
We shouldn't need this check if `loadModule` (or `processModule0`) helps us check whether a loaded script has been fully-cached every single time